src/state_circuit/state.rs

```diff
@@ -0,0 +1,896 @@
+    const ADDRESS_MAX: usize,
+    const ADDRESS_INCR: bool,
+    const ADDRESS_DIFF_STRICT: bool,
+> {
+    q_first: Selector,
+    q_not_first: Selector,
+    address: Column<Advice>,
+    address_diff_inv: Column<Advice>,
+    global_counter: Column<Advice>,
+    value: Column<Advice>,
+    flag: Column<Advice>,
+    global_counter_table: Column<Fixed>,
+    address_table_zero: Column<Fixed>,
+    is_zero: IsZeroConfig<F>,
+    mono_incr: MonotoneConfig,
+    _marker: PhantomData<F>,
+}
+
+impl<
+        F: FieldExt,
+        const MAX_MEMORY_ROWS: usize,
+        const GLOBAL_COUNTER_MAX: usize,
+        const ADDRESS_MAX: usize,
+        const ADDRESS_INCR: bool,
+        const ADDRESS_DIFF_STRICT: bool,
+    >
+    Config<F, MAX_MEMORY_ROWS, GLOBAL_COUNTER_MAX, ADDRESS_MAX, ADDRESS_INCR, ADDRESS_DIFF_STRICT>
+{
+    /// Set up custom gates and lookup arguments for this configuration.
+    pub(crate) fn configure(meta: &mut ConstraintSystem<F>) -> Self {
+        let q_first = meta.selector();
+        let q_not_first = meta.selector();
+        let address = meta.advice_column();
+        let address_diff_inv = meta.advice_column();
+        let global_counter = meta.advice_column();
+        let value = meta.advice_column();
+        let flag = meta.advice_column();
+        let global_counter_table = meta.fixed_column();
+        let address_table_zero = meta.fixed_column();
+
+        let is_zero = IsZeroChip::configure(
+            meta,
+            q_not_first,
+            |meta| {
+                let value_a = meta.query_advice(address, Rotation::cur());
+                let value_b = meta.query_advice(address, Rotation::prev());
+                value_a - value_b
+            },
+            address_diff_inv,
+        );
+
+        let mono_incr =
+            MonotoneChip::<F, ADDRESS_MAX, ADDRESS_INCR, ADDRESS_DIFF_STRICT>::configure(
+                meta,
+                |meta| meta.query_selector(q_not_first),
+                address,
+            );
+
+        // A gate for the first operation (does not need Rotation::prev).
+        meta.create_gate("First row operation", |meta| {
+            let q_first = meta.query_selector(q_first);
+
+            let value = meta.query_advice(value, Rotation::cur());
+            let flag = meta.query_advice(flag, Rotation::cur());
+            let global_counter = meta.query_advice(global_counter, Rotation::cur());
+
+            //      - values[0] == [0]
+            //      - flags[0] == 1
+            //      - global_counters[0] == 0
+
+            // TODO: global_counter can't be 0 for both - memory and stack init rows
+
+            let address = meta.query_advice(address, Rotation::cur());
+            let one = Expression::Constant(F::one());
+            let max_address = Expression::Constant(F::from_u64(ADDRESS_MAX as u64));
+            // For stack init operation: ADDRESS_INCR is false (thus 0) and the address must be ADDRESS_MAX (that is 1023)
```

Comment:
ADDRESS_INCR = 0 means decreasing, ADDRESS_INCR = 1 means decreasing. However, I talked with Han about this and the plan is to have stack and memory addresses both increasing. I planned to make these changes in #24 though because there will be some changes related to this anyway.